### ISSUE TITLE
Switch to dtype-independent buffers?

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#3275]
## Current situation

- Scipp uses a template, `scipp::core::element_array` to hold values of a `Variable`.
- Strides are given in units of the type, i.e., for a variable with `dtype=float` a stride of `1` mean 4 bytes.

### Broken functionality for some dtypes

There are a number of situations where operations are independent of the dtype, since all they do is move or copy elements. Prime example are binning or grouping operations. We have and are repeatedly running into cases where we forgot to add support for all relevant dtypes. The most recent example is #3271. This could have been avoided by operating on an underlying buffer of bytes (`char`?).

### Performance

`bin` could make use of some additional optimizations if buffers could be reused for different dtypes. There may also be other applications.

### Side issue

NumPy uses strides that are given in bytes. Currently we have to do a translation in the Python layer.

## Solution

The solution would be to use a buffer of bytes (`char`, or does modern C++ have something more suitable?), which can then be interpreted with different dtypes.

- All code that is handling strides would need updating.
- We need to come up with a solution for transforming data without using the dtype. Concretely, we would probably use the element size and some sort of memcopy in such cases.

**Comment by SimonHeybrock** [comment]
> > We need to come up with a solution for transforming data without using the dtype. Concretely, we would probably use the element size and some sort of memcopy in such cases.
> 
> Doesn't it need to resolve the dtype eventually? Otherwise, we cannot know how to, e.g., add two elements.

This is only about operations that move or copy elements such as `bin` or `where`. Other operations would operate in typed buffers as usual.

**Comment by jl-wynen** [comment]
So would we keep the existing `element_array<T>` but change it to sit on top of an untyped buffer? Or how would other operations access the type?

**Comment by SimonHeybrock** [comment]
> So would we keep the existing `element_array<T>` but change it to sit on top of an untyped buffer? Or how would other operations access the type?

Yes, this would be the rough idea, even though I have not though about how this would be handled under the hood.
